kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -425,29 +428,30 @@ void cacheRealmPermsChanged(String realmRef, Set<String> roles, Set<String> perm
 
         Set<Member> members = azg.getMembers();
         if (members != null && !members.isEmpty()) {
-            for (Member member : members) {
-                if (member != null && member.isActive() && member.getUserId() != null) {
-		    boolean canSwap = (member.getRole().isAllowed(SiteService.SITE_ROLE_SWAP));
-                    for (String perm : permissions) {
-                        if (perm != null) {
-                            keysToInvalidate.add(makeCacheKey(member.getUserId(), null, perm, azgRef, false));
+            for (String perm : permissions) {
+                if (perm != null) {
+                    HashSet<String> permKeysToInvalidate = new HashSet<>();
```

Comment:
I would consider allocating this hashset once outside the permissions loop and just clear it each iteration. The size should be the same each time so we should save on both allocations and resizing costs by just reusing the set.